backend/handler/filesystem/roms_handler.py

```diff
@@ -84,35 +165,106 @@ def _exclude_multi_roms(self, roms) -> list[str]:
 
         return [f for f in roms if f not in filtered_files]
 
-    def get_rom_files(self, rom: str, roms_path: str) -> list[str]:
-        rom_files: list[str] = []
-
-        for path, _, files in os.walk(f"{roms_path}/{rom}"):
-            for f in self._exclude_files(files, "multi_parts"):
-                rom_files.append(f"{Path(path, f)}".replace(f"{roms_path}/{rom}/", ""))
+    def _calculate_rom_hashes(self, file_path: Path) -> dict[str, str]:
+        mime = magic.Magic(mime=True)
+        file_type = mime.from_file(file_path)
+        extension = Path(file_path).suffix.lower()
+
+        crc_c = 0
+        md5_h = hashlib.md5(usedforsecurity=False)
+        sha1_h = hashlib.sha1(usedforsecurity=False)
+
+        def update_hashes(chunk: bytes):
+            md5_h.update(chunk)
+            sha1_h.update(chunk)
+            nonlocal crc_c
+            crc_c = binascii.crc32(chunk, crc_c)
+
+        if extension == ".zip" or file_type == "application/zip":
+            for chunk in read_zip_file(file_path):
+                update_hashes(chunk)
+
+        elif extension == ".tar" or file_type == "application/x-tar":
+            for chunk in read_tar_file(file_path):
+                update_hashes(chunk)
+
+        elif extension == ".gz" or file_type == "application/x-gzip":
+            for chunk in read_gz_file(file_path):
+                update_hashes(chunk)
+
+        elif extension == ".7z" or file_type == "application/x-7z-compressed":
+            for chunk in read_7z_file(file_path):
+                update_hashes(chunk)
+
+        elif extension == ".bz2" or file_type == "application/x-bzip2":
+            for chunk in read_bz2_file(file_path):
+                update_hashes(chunk)
+
+        else:
+            with open(file_path, "rb") as f:
+                # Read in chunks to avoid memory issues
+                while chunk := f.read(FILE_READ_CHUNK_SIZE):
+                    update_hashes(chunk)
+
+        return {
+            "crc_hash": (crc_c & 0xFFFFFFFF).to_bytes(4, byteorder="big").hex(),
+            "md5_hash": md5_h.hexdigest(),
+            "sha1_hash": sha1_h.hexdigest(),
+        }
+
+    def get_rom_files(self, rom: str, roms_path: str) -> list[RomFile]:
```

Comment:
Comparing this implementation with the old one, I think we need the call to `_exclude_files`, to honor any exclusion the user set.